src/classes/CW_AudienceGroupTriggerHandler.cls

```diff
@@ -0,0 +1,219 @@
+		for(CW_Audience_Group__c ag : newTriggerMap.values())
+		{
+			if(ag.Is_Control_Group__c != oldTriggerMap.get(ag.Id).Is_Control_Group__c
+				|| ag.Control_Group__c != oldTriggerMap.get(ag.Id).Control_Group__c)
+			{
+				changedGroups.add(ag);
+			}
+		}
+		
+		changedGroups = audienceGroupsWithAGS(changedGroups);
+		List<CW_Audience_Group_Schedule__c> agss = new List<CW_Audience_Group_Schedule__c>();
+		for(CW_Audience_Group__c ag : changedGroups)
+		{
+			for(CW_Audience_Group_Schedule__c ags : ag.Audience_Group_Schedules__r)
+			{
+				ags.Is_Control_Group__c = ag.Is_Control_Group__c;
+				ags.Control_Group_Schedule__c = null;	// reset
+			}
+			agss.addAll(ag.Audience_Group_Schedules__r);
+			ag.Audience_Group_Schedules__r.clear();		// reduce heap size
+		}
+		
+		matchControlGroups(agss, new Map<Id,CW_Audience_Group__c>(changedGroups));
+	}
+	
+	// create audience schedule group on after insert
+	public static void createAGS(Map<Id,CW_Audience_Group__c> groupsMap)
+	{
+		
+		// part 1: create AGS
+		Map<Id,List<CW_Schedule__c>> campSchedules = new  Map<Id,List<CW_Schedule__c>>(); // k: campaign id, v: list of schedules of camp
+		for(CW_Audience_Group__c gr : groupsMap.values())
+		{
+			campSchedules.put(gr.Campaign__c,new List<CW_Schedule__c>());
+		}
+		
+		for(CW_Campaign__c camp : [select id, (select id from Schedules__r) from CW_Campaign__c where id in :campSchedules.keySet()])
+		{
+			campSchedules.get(camp.Id).addAll(camp.Schedules__r);
+		}
+		
+		// now that we have all schedules, create AGS - only after insert
+		List<CW_Audience_Group_Schedule__c> agss = new List<CW_Audience_Group_Schedule__c>();
+		for(CW_Audience_Group__c gr : groupsMap.values())
+		{
+			for(CW_Schedule__c sch : campSchedules.get(gr.Campaign__c))
+			{
+				CW_Audience_Group_Schedule__c ags = new CW_Audience_Group_Schedule__c();
+				ags.Audience_Group__c = gr.id;
+				ags.Schedule__c = sch.id;
+				ags.Is_Control_Group__c = gr.Is_Control_Group__c;
+				agss.add(ags);
+			}
+		}
+		insert agss;
+		
+		// part2: match control group schedules
+		matchControlGroups(agss,groupsMap);
+	}
+	
+	// fix links to control groups
+	public static void matchControlGroups(List<CW_Audience_Group_Schedule__c> agss, Map<Id,CW_Audience_Group__c> groupsMap)
+	{
+		
+		Map<String,CW_Audience_Group_Schedule__c> allAGS = createComboSet(agss);
+		
+		Set<Id> missingIds = new Set<Id>();	// control groups schedules we will need to select from the database
+		for(CW_Audience_Group__c gr : groupsMap.values())
+		{
+			if(gr.Control_Group__c != null && !groupsMap.containsKey(gr.Control_Group__c))	// if the control group is not in trigger list
+			{
+				missingIds.add(gr.Control_Group__c);
+			}
+		}
+		
+		if(!missingIds.isEmpty())	// find any that might be missing from the trigger list
+		{
+			for(CW_Audience_Group_Schedule__c ags : 
+					[select id, Audience_Group__c, Schedule__c from CW_Audience_Group_Schedule__c where Audience_Group__c in :missingIds])
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.